opm/ml/ml_model.hpp

```diff
@@ -0,0 +1,379 @@
+/*
+  Copyright (c) 2016 Robert W. Rose
+  Copyright (c) 2018 Paul Maevskikh
+  Copyright (c) 2024 NORCE
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef ML_MODEL_H_
+#define ML_MODEL_H_
+
+#include <algorithm>
+#include <chrono>
+#include <cmath>
+#include <cstdio>
+#include <numeric>
+#include <string>
+#include <vector>
+
+#include <opm/material/densead/Math.hpp>
+#include <opm/common/ErrorMacros.hpp>
+#include <fmt/format.h>
+
+
+namespace Opm {
+
+// NN layer
```

Comment:
This comment seems misplaced?

Although, it made me think that you should put all of these classes and functions inside a subnamespace (for example `NN`), you can do this by changing the above to:
```c++
namespace Opm::NN
```
And then you can remove the NN from all the class names below, making the rest easier to read and understand. Since you put all of this in a subdir `ml` maybe `ML` better than `NN` as namespace name though.